{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: genbank in /opt/anaconda3/lib/python3.9/site-packages (0.110)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: optipyzer in /opt/anaconda3/lib/python3.9/site-packages (0.4.1)\n",
      "Requirement already satisfied: pydantic in /opt/anaconda3/lib/python3.9/site-packages (from optipyzer) (1.10.7)\n",
      "Requirement already satisfied: requests in /opt/anaconda3/lib/python3.9/site-packages (from optipyzer) (2.28.1)\n",
      "Requirement already satisfied: typing-extensions>=4.2.0 in /opt/anaconda3/lib/python3.9/site-packages (from pydantic->optipyzer) (4.3.0)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /opt/anaconda3/lib/python3.9/site-packages (from requests->optipyzer) (2.0.7)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/anaconda3/lib/python3.9/site-packages (from requests->optipyzer) (3.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /opt/anaconda3/lib/python3.9/site-packages (from requests->optipyzer) (1.26.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/anaconda3/lib/python3.9/site-packages (from requests->optipyzer) (2021.10.8)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install optipyzer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in antigens\n",
    "import pandas as pd\n",
    "\n",
    "antigens = pd.read_csv('../antigens/antigens.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein ID</th>\n",
       "      <th>Protein Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Smp_075800.1</td>\n",
       "      <td>MMLFSLFLISILHILLVKCQLDTNYEVSDETVSDNNKWAVLVAGSN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Smp_179950.1</td>\n",
       "      <td>MLISVLYIASLISHLEAHISIKNEKFEPLSDDIISYINEHPNAGWR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Smp_158420.1</td>\n",
       "      <td>MLISVLCIASLITHLEAHISIKNEKFEPLSHDIISYINKHLDARRE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MEG4.161011091</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MEG4.1610110</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MEG4.11101100</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MEG4.1700100</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MEG4.1200001</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MEG4.11711010172</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNKIFQTDKFHLKEYI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MEG4.1700110</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MEG4.11601101172</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKIFQTDKF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MEG4.11710010172</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNKLKEYIEDKNVDIR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MEG4.1901110</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MEG4.1610100</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MEG4.1711010</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MEG4.1701100</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MEG4.11500110</td>\n",
       "      <td>MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTIKFSRLINSM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MEG8.2400100</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MEG8.21500100</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MEG8.21501010</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MEG8.21510010181</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MEG8.21401000</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MEG8.21200110</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MEG8.21411000</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MEG8.21510100</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MEG8.21200100</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MEG8.2511000</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MEG8.2901100</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MEG8.2511010</td>\n",
       "      <td>MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MEG3.1711000</td>\n",
       "      <td>MLFFALILIISLHSFDCVFTARETRDAALSCKKHCEGNNEYVTRYC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MEG3.1710010</td>\n",
       "      <td>MLFFALILIISLHSFDCVFTARETRDAALSCKKHCEGNNEYVTRYC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MEG15410110</td>\n",
       "      <td>MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MEG151001110</td>\n",
       "      <td>MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MEG15400100</td>\n",
       "      <td>MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Protein ID                                   Protein Sequence\n",
       "0       Smp_075800.1  MMLFSLFLISILHILLVKCQLDTNYEVSDETVSDNNKWAVLVAGSN...\n",
       "2       Smp_179950.1  MLISVLYIASLISHLEAHISIKNEKFEPLSDDIISYINEHPNAGWR...\n",
       "3       Smp_158420.1  MLISVLCIASLITHLEAHISIKNEKFEPLSHDIISYINKHLDARRE...\n",
       "4     MEG4.161011091  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "5       MEG4.1610110  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "6      MEG4.11101100  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "7       MEG4.1700100  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "8       MEG4.1200001  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "9   MEG4.11711010172  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNKIFQTDKFHLKEYI...\n",
       "10      MEG4.1700110  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "11  MEG4.11601101172  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKIFQTDKF...\n",
       "12  MEG4.11710010172  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNKLKEYIEDKNVDIR...\n",
       "13      MEG4.1901110  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "14      MEG4.1610100  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "15      MEG4.1711010  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "16      MEG4.1701100  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEE...\n",
       "17     MEG4.11500110  MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTIKFSRLINSM...\n",
       "18      MEG8.2400100  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "19     MEG8.21500100  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "20     MEG8.21501010  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "21  MEG8.21510010181  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "22     MEG8.21401000  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "23     MEG8.21200110  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "24     MEG8.21411000  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "25     MEG8.21510100  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "26     MEG8.21200100  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "27      MEG8.2511000  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "28      MEG8.2901100  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "29      MEG8.2511010  MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLT...\n",
       "30      MEG3.1711000  MLFFALILIISLHSFDCVFTARETRDAALSCKKHCEGNNEYVTRYC...\n",
       "31      MEG3.1710010  MLFFALILIISLHSFDCVFTARETRDAALSCKKHCEGNNEYVTRYC...\n",
       "32       MEG15410110  MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL...\n",
       "33      MEG151001110  MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL...\n",
       "35       MEG15400100  MLNRFIVILVFVFVGIVTFDNVQGQRDPPRTNNTITHTTNHYVGKL..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get each unique protein\n",
    "protiens = antigens.drop_duplicates(subset='Protein ID')[[\"Protein ID\", \"Protein Sequence\"]]\n",
    "protiens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.9/site-packages/Bio/SeqFeature.py:1040: BiopythonParserWarning: Attempting to fix invalid location '4907..203' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGG...AAA'), id='SpyTag-Antigen-H6_Templ', name='SpyTag-Antigen-H6_Templ', description='', dbxrefs=[])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO, SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation, ExactPosition\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Read the GenBank template file\n",
    "record: SeqRecord = SeqIO.read('SpyTag-Antigen-H6 Template.gb', 'genbank')\n",
    "record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExactPosition(3236)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find the location to insert\n",
    "for feature in record.features:\n",
    "    if 'label' in feature.qualifiers:\n",
    "        if feature.qualifiers['label'][0] == \"Linker\":\n",
    "            insert_location: ExactPosition = feature.location.end  # End of \"Linker\"\n",
    "\n",
    "assert insert_location is not None \n",
    "\n",
    "insert_location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import optipyzer\n",
    "\n",
    "api = optipyzer.API()\n",
    "\n",
    "def optimize(protein: str) -> str:\n",
    "    # Perform codon optimization for the protein and return the optimized DNA sequence\n",
    "    optimization = api.optimize(\n",
    "        seq=protein,\n",
    "        seq_type='protein',\n",
    "        weights={'e_coli': 1.0}, # We are expressing in E. coli\n",
    "    )\n",
    "    return optimization['optimized_sd']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.9/site-packages/Bio/SeqFeature.py:1040: BiopythonParserWarning: Attempting to fix invalid location '4907..203' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "for protien_id, protien_seq in protiens.to_numpy():\n",
    "    # Run codon optimization\n",
    "    nucleotide_seq = optimize(protien_seq)[:-3] # Remove the stop codon\n",
    "    \n",
    "    # Insert the protein sequence at the correct location\n",
    "    new_seq = record.seq[:insert_location] + Seq(nucleotide_seq) + record.seq[insert_location:]\n",
    "\n",
    "    # Update the sequence in the record\n",
    "    record.seq = new_seq\n",
    "\n",
    "    # Calculate the length of the inserted sequence\n",
    "    insert_length = len(nucleotide_seq)\n",
    "\n",
    "    # Update features after the insertion\n",
    "    updated_features = []\n",
    "    for feature in record.features:\n",
    "        if feature.location.start >= insert_location:\n",
    "            # If the feature starts after the insertion point, shift its location\n",
    "            new_start = feature.location.start + insert_length\n",
    "            new_end = feature.location.end + insert_length\n",
    "            new_feature = SeqFeature(FeatureLocation(new_start, new_end), type=feature.type, qualifiers=feature.qualifiers)\n",
    "            updated_features.append(new_feature)\n",
    "        elif feature.type == \"CDS\":   \n",
    "            # Update the CDS to include the inserted sequence\n",
    "            new_start = feature.location.start\n",
    "            new_end = feature.location.end + insert_length\n",
    "            feature.qualifiers.pop('translation', None) # let the translation be recalculated\n",
    "            new_feature = SeqFeature(FeatureLocation(new_start, new_end), type=feature.type, qualifiers=feature.qualifiers)\n",
    "            updated_features.append(new_feature)\n",
    "        else:\n",
    "            # Keep the feature as is if it occurs before the insertion point\n",
    "            updated_features.append(feature)\n",
    "\n",
    "    # Add the new feature for the inserted sequence\n",
    "    inserted_feature = SeqFeature(\n",
    "        FeatureLocation(insert_location, insert_location + insert_length),\n",
    "        type=\"misc_feature\",\n",
    "        qualifiers={\n",
    "            \"label\": [protien_id],\n",
    "            \"ApEinfo_revcolor\": \"#ff0000\",  # Red color for the inserted antigen\n",
    "            \"ApEinfo_fwdcolor\": \"#ff0000\"\n",
    "        }\n",
    "    )\n",
    "    updated_features.append(inserted_feature)\n",
    "\n",
    "    # Replace the old features with the updated ones\n",
    "    record.features = updated_features\n",
    "\n",
    "    # Write the modified record back to a new GenBank file\n",
    "    construct_name = f'SpyTag-{protien_id}-H6'\n",
    "    filename = f'{construct_name}.gb'\n",
    "    with open(filename, 'w') as output_file:\n",
    "        SeqIO.write(record, output_file, 'genbank')\n",
    "        \n",
    "    with open(filename, \"r\") as file:\n",
    "        updated_file = file.read().replace(\"SpyTag-Antigen-H6_Templ\", construct_name)\n",
    "    \n",
    "    with open(filename, \"w\") as file:\n",
    "        file.write(updated_file)\n",
    "        \n",
    "    # Reset the record for the next iteration\n",
    "    record = SeqIO.read('SpyTag-Antigen-H6 Template.gb', 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
